core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java

```diff
@@ -100,6 +106,53 @@ public static EnumerableHashJoin create(
         condition, variablesSet, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null || collation == RelCollations.EMPTY) {
+      return null;
+    }
+
+    List<Integer> requiredKeys = RelCollations.ordinals(collation);
+    ImmutableBitSet requiredKeySet = ImmutableBitSet.of(requiredKeys);
+
+    ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
+    ImmutableBitSet rightKeySet = ImmutableBitSet.of(joinInfo.rightKeys)
+        .shift(left.getRowType().getFieldCount());
+
+    // HashJoin traits passdown shall only consider left/right outer join.
+    // It is because for a hash-based implementation, only non-hashed side can
+    // preserve ordering, thus only for left/right outer join, we are sure which
+    // side is non-hashed side (the outer child).
+    if (joinType == JoinRelType.LEFT) {
+      // Only consider exact key match for now
+      if (requiredKeySet.equals(leftKeySet)) {
+        return Pair.of(
+            required,
+            ImmutableList.of(
+                required,
+                required.replace(RelCollations.EMPTY)));
+      }
+    } else if (joinType == JoinRelType.RIGHT) {
+      // Only consider exact key match for now
+      if (requiredKeySet.equals(rightKeySet)) {
+        RelCollation rightCollation = RelCollations.shift(collation,
+            -left.getRowType().getFieldCount());
+        return Pair.of(
+            required,
+            ImmutableList.of(
+                required.replace(RelCollations.EMPTY),
+                required.replace(rightCollation)));
+      }
+    }
+
+    return null;
+  }
+
+//  @Override public DeriveMode getDeriveMode() {
+//    return DeriveMode.BOTH;
```

Comment:
should prohibit for full outer join.

Reply:
Sounds good for full outer join (will use `DeriveMode.PROHIBITED` to dis-allow it)

Still understanding why RIGHT OUTER JOIN should be forbidon (why EnumerableHashJoin always builds hash table on right input).

Reply:
I am a little bit skeptical about it. It shouldn't work that way. If there isn't a match, it should generate nulls on left, but building hash table on right side will make it impossible. Can you write a query and execute it?

Reply:
Will do.

Reply:
I confirmed that indeed EnumerableHashJoin always builds hash table on right side. 

For right/full outer join, to make result right, the implementation keeps a set of keys from right side. During probing it also removes keys from that set. At the end it will know how many keys from right side does not have a match with left side, thus it can produces null with right tuples. 


So I will prohibit traits propogation for right/full outer join.

Reply:
Cool, thanks for checking.